src/checkers/inference/solver/GeneralSolver.java

```diff
@@ -0,0 +1,407 @@
+            this.useGraph = true;
+        } else {
+            this.useGraph = false;
+        }
+
+        if (backEndType.equals(BackEndType.LOGIQL)) {
+            // Configure solving strategy.
+            this.solveInParallel = false;
+        } else if (solveInParallel == null || solveInParallel.equals(Constants.TRUE)) {
+            this.solveInParallel = true;
+        } else {
+            this.solveInParallel = false;
+        }
+
+        if (collectStatistic == null || collectStatistic.equals(Constants.FALSE)) {
+            // Configure statistic collection.
+            this.collectStatistic = false;
+        } else if (collectStatistic.equals(Constants.TRUE)) {
+            this.collectStatistic = true;
+        }
+
+        // Sanitize the configuration if it needs.
+        sanitizeConfiguration();
+        System.out.println("Configuration: \nback end type: " + this.backEndType.simpleName + "; \nuseGraph: "
+                + this.useGraph + "; \nsolveInParallel: " + this.solveInParallel + ".");
+    }
+
+    protected void configureLattice(QualifierHierarchy qualHierarchy) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        lattice = latticeBuilder.buildLattice(qualHierarchy);
+    }
+
+    protected TwoQualifiersLattice createTwoQualifierLattice(AnnotationMirror top, AnnotationMirror bottom) {
+        LatticeBuilder latticeBuilder = new LatticeBuilder();
+        return latticeBuilder.buildTwoTypeLattice(top, bottom);
+    }
+
+    /**
+     * This method creates a ConstraintSerializer, which can deliver the
+     * constraint/slot to the real serializer. If a costume serialization logic
+     * is needed, user can have a subclass of ConstraintSerializer, override the
+     * serialize method with the costume logic, and then override this method
+     * with returning the instance of new subclass.
+     * 
+     * @param backEndType
+     * @param lattice
+     * @return A deliverer Serializer.
+     */
+    protected Serializer<?, ?> createSerializer(BackEndType backEndType, Lattice lattice) {
+        return new ConstraintSerializer<>(backEndType, lattice);
+    }
+
+    protected ConstraintGraph generateGraph(Collection<Slot> slots, Collection<Constraint> constraints,
+            ProcessingEnvironment processingEnvironment) {
+        GraphBuilder graphBuilder = new GraphBuilder(slots, constraints);
+        ConstraintGraph constraintGraph = graphBuilder.buildGraph();
+        return constraintGraph;
+    }
+
+    protected BackEnd<?, ?> createBackEnd(BackEndType backEndType, Map<String, String> configuration,
+            Collection<Slot> slots, Collection<Constraint> constraints,
+            QualifierHierarchy qualHierarchy, ProcessingEnvironment processingEnvironment,
+            Lattice lattice, Serializer<?, ?> defaultSerializer) {
+
+        BackEnd<?, ?> backEnd = null;
+
+        try {
+            Class<?> backEndClass = Class.forName(backEndType.fullyQualifiedName + "BackEnd");
```

Comment:
Hi @Jianchu , just some thought on loading backend and serializer instance:

Could we move the loading instance logic to `BackEndType`? E.g.

```java

enum BackEndType {
   MAX_SAT("maxsat", MaxSatBackEnd.class, MaxSatSerializer.class),
   ....;
   public final String simpleName;
   private final Class<? extends BackEnd<?, ?>> backEndClass;
   private final Class<? extends Serializer<?, ?>> defaultSerializerClass;

   private BackEndType(String simpleName, Class<? extends BackEnd<?, ?>> backEndClass, Class<? extends Serialzier<?, ?>> serializerClass) {
 //initialize fields
}

public BackEnd<?, ?> getNewBackEndInstance( /*passing all needed argument here*/) {
 // move the reflection logic here to return a new instance
}

public Serializer<?, ?> getNewDefaultSerializerInstance(/*passing all need argument here*/) {
  // move the reflection logic here to return a new instance
}
```

Reason of this proposing:

1. Using string to controll `fullyQualifiedName` isn't robust --- if BackEnd path changed, this string won't auto-updated by IDE and may crash.

2. We can remove the `ConstraintSerializer` class, this class is seems just a delegator that controls how to load a serializer for BackEnd, which seems is part of the configuration of a BackEnd. Then moving this logic to `BackEndType` can centralized the configuration logic into one place.

What do you think?